File: mcp-server/src/autoagent_mcp/tree_cache.py
"""In-memory snapshot cache for UI tree delta comparisons (TASK-0405C).

:class:`TreeCache` stores numbered snapshots of ``dump_tree`` results and
computes diffs between them so :func:`dump_tree_delta` can return only the
nodes that changed since a previous call.

Why not store raw engine output?
    The cache operates on the *filtered* node list that the agent already
    sees (after ``include_invisible`` / ``max_depth`` are applied), so the
    delta is always meaningful from the agent's perspective.

Snapshot IDs
    8-character hex strings (4 bytes of randomness).  They are opaque — the
    agent passes them back as-is.  If a snapshot has been evicted (the cache
    holds at most *max_snapshots* entries, oldest evicted first), the delta
    call falls back to returning the full tree.

Thread safety
    All mutations are protected by a :class:`threading.Lock`.

Usage::

    from autoagent_mcp.tree_cache import get_tree_cache, reset_tree_cache

    cache = get_tree_cache()
    snap_id = cache.store(nodes)           # returns "a1b2c3d4"
    delta   = cache.diff(snap_id, nodes2)  # DeltaResult or None
"""

from __future__ import annotations

import json
import secrets
import threading
from dataclasses import dataclass, field

# Maximum number of snapshots kept in memory.  Oldest entry is evicted when
# the limit is exceeded.
_DEFAULT_MAX_SNAPSHOTS: int = 20


# ---------------------------------------------------------------------------
# Public types
# ---------------------------------------------------------------------------


@dataclass
class DeltaResult:
    """Result of :meth:`TreeCache.diff`.

    Attributes:
        snapshot_id:     ID of the *new* snapshot (use for the next delta call).
        changed:         Nodes that were **added or modified** since the baseline
                         (full node data, same schema as ``dump_tree`` nodes).
        removed_ids:     IDs that were present in the baseline but are gone now.
        unchanged_count: Number of nodes that did not change (informational).
        full_snapshot:   ``True`` when the baseline snapshot was not found and
                         the result contains the complete tree instead of a diff.
    """

    snapshot_id: str
    changed: list[dict] = field(default_factory=list)
    removed_ids: list[str] = field(default_factory=list)
    unchanged_count: int = 0
    full_snapshot: bool = False


# ---------------------------------------------------------------------------
# Cache implementation
# ---------------------------------------------------------------------------


class TreeCache:
    """Snapshot store with LRU eviction.

    Args:
        max_snapshots: Maximum number of snapshots to keep (default 20).
    """

    def __init__(self, max_snapshots: int = _DEFAULT_MAX_SNAPSHOTS) -> None:
        self._max = max_snapshots
        self._lock = threading.Lock()
        # snapshot_id → {node_id: canonical_json}
        self._snapshots: dict[str, dict[str, str]] = {}
        self._order: list[str] = []  # insertion order for eviction

    # ---- public API -------------------------------------------------------

    def store(self, nodes: list[dict]) -> str:
        """Persist *nodes* as a new snapshot and return its ID."""
        snap_id = secrets.token_hex(4)  # 8 hex chars, e.g. "a1b2c3d4"
        mapping = {n["id"]: _canonical(n) for n in nodes if "id" in n}
        with self._lock:
            self._snapshots[snap_id] = mapping
            self._order.append(snap_id)
            self._evict()
        return snap_id

    def diff(self, since_id: str, new_nodes: list[dict]) -> DeltaResult:
        """Compute the delta from snapshot *since_id* to *new_nodes*.

        If *since_id* is unknown (expired or invalid), the result is treated
        as a fresh full snapshot (``full_snapshot=True``).

        Always stores *new_nodes* as a new snapshot and returns its ID.
        """
        with self._lock:
            baseline: dict[str, str] | None = self._snapshots.get(since_id)

        if baseline is None:
            # Baseline expired or never existed — return the full tree.
            new_snap_id = self.store(new_nodes)
            return DeltaResult(
                snapshot_id=new_snap_id,
                changed=list(new_nodes),
                removed_ids=[],
                unchanged_count=0,
                full_snapshot=True,
            )

        # Build new mapping for comparison.
        new_map: dict[str, tuple[dict, str]] = {
            n["id"]: (n, _canonical(n)) for n in new_nodes if "id" in n
        }

        changed: list[dict] = []
        unchanged_count: int = 0
        for node_id, (node, canon) in new_map.items():
            if node_id not in baseline or baseline[node_id] != canon:
                changed.append(node)
            else:
                unchanged_count += 1

        removed_ids = [nid for nid in baseline if nid not in new_map]

        new_snap_id = self.store(new_nodes)
        return DeltaResult(
            snapshot_id=new_snap_id,
            changed=changed,
            removed_ids=removed_ids,
            unchanged_count=unchanged_count,
            full_snapshot=False,
        )

    def clear(self) -> None:
        """Remove all snapshots."""
        with self._lock:
            self._snapshots.clear()
            self._order.clear()

    def snapshot_count(self) -> int:
        """Return the number of snapshots currently stored."""
        with self._lock:
            return len(self._snapshots)

    # ---- private ----------------------------------------------------------

    def _evict(self) -> None:
        """Evict oldest entries until at or below *_max*. Caller holds lock."""
        while len(self._order) > self._max:
            old_id = self._order.pop(0)
            self._snapshots.pop(old_id, None)


# ---------------------------------------------------------------------------
# Private helpers
# ---------------------------------------------------------------------------


def _canonical(node: dict) -> str:
    """Return a stable, compact JSON string for node equality comparison."""
    return json.dumps(node, sort_keys=True, separators=(",", ":"),
                      ensure_ascii=False)


# ---------------------------------------------------------------------------
# Module-level singleton
# ---------------------------------------------------------------------------

_cache = TreeCache()


def get_tree_cache() -> TreeCache:
    """Return the process-level :class:`TreeCache` singleton."""
    return _cache


def reset_tree_cache() -> None:
    """Clear all snapshots (e.g. on engine reconnect or test teardown)."""
    _cache.clear()

File: adapters/unity/Assets/AutoAgent/Runtime/Server/TreeCache.cs
using System;
using System.Collections.Generic;
using System.Text;
using UnityEngine;

namespace AutoAgent
{
    /// <summary>
    /// Result returned by <see cref="TreeCache.Diff"/>.
    /// </summary>
    internal struct DeltaResult
    {
        public string      snapshotId;
        public List<NodeData> changed;
        public List<string>   removedIds;
        public int            unchangedCount;
        public bool           fullSnapshot;

        public static DeltaResult Full(List<NodeData> nodes, string snapshotId) =>
            new DeltaResult
            {
                snapshotId     = snapshotId,
                changed        = nodes,
                removedIds     = new List<string>(),
                unchangedCount = 0,
                fullSnapshot   = true,
            };
    }

    /// <summary>
    /// In-memory snapshot cache that computes deltas between two
    /// <c>dump_tree</c> calls, returning only added / modified / removed
    /// nodes.
    ///
    /// <para>Thread-safe: all public methods acquire a lock internally.</para>
    /// <para>Keeps at most <c>MaxSnapshots</c> entries (default 20),
    /// evicting the oldest on overflow.</para>
    /// </summary>
    internal static class TreeCache
    {
        const int DefaultMaxSnapshots = 20;

        static readonly object _lock = new object();
        static readonly Dictionary<string, Dictionary<string, string>> _snapshots =
            new Dictionary<string, Dictionary<string, string>>();
        static readonly List<string> _order = new List<string>();
        static int _maxSnapshots = DefaultMaxSnapshots;

        // ------------------------------------------------------------------
        // Public API
        // ------------------------------------------------------------------

        /// <summary>
        /// Store a full node list as a new snapshot.  Returns an 8-char hex
        /// snapshot ID for use in a subsequent <see cref="Diff"/> call.
        /// </summary>
        public static string Store(List<NodeData> nodes)
        {
            string snapId = GenerateId();
            var mapping = new Dictionary<string, string>();
            foreach (var n in nodes)
            {
                if (string.IsNullOrEmpty(n.Id)) continue;
                mapping[n.Id] = ComputeNodeHash(n);
            }
            lock (_lock)
            {
                _snapshots[snapId] = mapping;
                _order.Add(snapId);
                EvictLocked();
            }
            return snapId;
        }

        /// <summary>
        /// Compute the delta between the baseline snapshot *sinceId* and
        /// *newNodes*.  Always stores *newNodes* as a fresh snapshot.
        ///
        /// When *sinceId* is unknown (expired / never existed) the result
        /// has <see cref="DeltaResult.fullSnapshot"/> == true.
        /// </summary>
        public static DeltaResult Diff(string sinceId, List<NodeData> newNodes)
        {
            Dictionary<string, string> baseline;
            lock (_lock) { _snapshots.TryGetValue(sinceId ?? "", out baseline); }

            if (baseline == null)
            {
                string snapId = Store(newNodes);
                return DeltaResult.Full(newNodes, snapId);
            }

            // Build new mapping.
            var newMap = new Dictionary<string, (NodeData node, string hash)>();
            foreach (var n in newNodes)
            {
                if (string.IsNullOrEmpty(n.Id)) continue;
                newMap[n.Id] = (n, ComputeNodeHash(n));
            }

            var changed = new List<NodeData>();
            int unchanged = 0;
            foreach (var (nid, (node, hash)) in newMap)
            {
                if (!baseline.TryGetValue(nid, out var oldHash) || oldHash != hash)
                    changed.Add(node);
                else
                    unchanged++;
            }

            var removed = new List<string>();
            foreach (var nid in baseline.Keys)
                if (!newMap.ContainsKey(nid))
                    removed.Add(nid);

            string newId = Store(newNodes);
            return new DeltaResult
            {
                snapshotId     = newId,
                changed        = changed,
                removedIds     = removed,
                unchangedCount = unchanged,
                fullSnapshot   = false,
            };
        }

        /// <summary>Remove all cached snapshots.</summary>
        public static void Clear()
        {
            lock (_lock) { _snapshots.Clear(); _order.Clear(); }
        }

        /// <summary>Number of snapshots currently stored.</summary>
        public static int SnapshotCount { get { lock (_lock) return _snapshots.Count; } }

        // ------------------------------------------------------------------
        // Internals
        // ------------------------------------------------------------------

        static readonly System.Random _rng = new System.Random();

        static string GenerateId()
        {
            // 4 random bytes → 8 hex chars, fast and sufficient for cache keys.
            int val;
            lock (_rng) { val = _rng.Next(); }
            return val.ToString("x8");
        }

        static string ComputeNodeHash(NodeData node)
        {
            var sb = new StringBuilder(256);
            AppendField(sb, node.Id);
            AppendField(sb, node.Type);
            AppendField(sb, node.EngineType);
            AppendField(sb, node.ParentId ?? "");
            AppendField(sb, string.Join(",", node.ChildrenIds));
            AppendField(sb, node.StableIdSource ?? "");

            var v = node.Visual;
            sb.Append(v.Position?[0] ?? 0); sb.Append(','); sb.Append(v.Position?[1] ?? 0); sb.Append('|');
            sb.Append(v.Size?[0] ?? 0); sb.Append(','); sb.Append(v.Size?[1] ?? 0); sb.Append('|');
            sb.Append(v.Anchor?[0] ?? 0); sb.Append(','); sb.Append(v.Anchor?[1] ?? 0); sb.Append('|');
            sb.Append(v.Visible); sb.Append('|');
            sb.Append(v.Alpha); sb.Append('|');
            AppendField(sb, v.Color ?? "");
            AppendField(sb, v.SpriteRef ?? "");
            if (v.WorldBounds != null && v.WorldBounds.Length == 4)
            { sb.Append(v.WorldBounds[0]); sb.Append(','); sb.Append(v.WorldBounds[1]);
              sb.Append(','); sb.Append(v.WorldBounds[2]); sb.Append(','); sb.Append(v.WorldBounds[3]); }
            sb.Append('|'); sb.Append(v.ZOrder); sb.Append('|');

            var b = node.Behavior;
            sb.Append(b.Interactable); sb.Append('|');
            sb.Append(b.RaycastTarget); sb.Append('|');
            AppendField(sb, string.Join(",", b.EventHandlers));
            AppendField(sb, string.Join(",", b.CustomScripts));
            AppendField(sb, string.Join(",", b.AttachedComponents));

            var m = node.Meta;
            if (m != null)
            {
                AppendField(sb, m.LogicalRole ?? "");
                AppendField(sb, m.Intent ?? "");
                AppendField(sb, string.Join(",", m.Tags ?? new List<string>()));
                if (m.StateSprites != null)
                    foreach (var kv in m.StateSprites)
                    { sb.Append(kv.Key); sb.Append('='); sb.Append(kv.Value); sb.Append(';'); }
            }

            // FNV-1a 64-bit: fast non-crypto hash, good distribution for content comparison.
            return Fnv1a64(sb.ToString()).ToString("x16");
        }

        static void AppendField(StringBuilder sb, string val)
        {
            sb.Append(val); sb.Append('|');
        }

        static ulong Fnv1a64(string s)
        {
            const ulong offset = 14695981039346656037UL;
            const ulong prime  = 1099511628211UL;
            ulong hash = offset;
            foreach (char c in s)
            {
                hash ^= (byte)(c & 0xFF);
                hash *= prime;
                hash ^= (byte)(c >> 8);
                hash *= prime;
            }
            return hash;
        }

        static void EvictLocked()
        {
            while (_order.Count > _maxSnapshots)
            {
                string old = _order[0];
                _order.RemoveAt(0);
                _snapshots.Remove(old);
            }
        }
    }
}
